src/cephadm/cephadmlib/container_types.py

```diff
@@ -606,3 +608,62 @@ def extract_uid_gid(
         raise Error(f'Failed to extract uid/gid for path {ex[0]}: {ex[1]}')
 
     raise RuntimeError('uid/gid not found')
+
+
+@functools.lru_cache()
+def _opt_key(value: str) -> str:
+    """Return a (long) option stripped of its value."""
+    return value.split('=', 1)[0]
+
+
+def _replace_container_arg(args: List[str], new_arg: str) -> None:
+    """Remove and replace arguments that have the same `--xyz` part as
+    the given `new_arg`. If new_arg is expected to have a value it
+    must be part of the new_arg string following an equal sign (`=`).
+    The existing arg may be a single or two strings in the input list.
+    """
+    key = _opt_key(new_arg)
+    has_value = key != new_arg
+    try:
+        idx = [_opt_key(v) for v in args].index(key)
+        if '=' in args[idx] or not has_value:
+            del args[idx]
+        else:
+            del args[idx]
+            del args[idx]
```

Comment:
Consider a list representing arguments like so:
a1 = ['--foo', '--speed=88', '--location=home', '--enable-flux']

Let's say we want to remove the --speed argument. If so we need to remove it's value too. idx first gets set to the position of an arg starting with `--speed` if speed is followed by the `=` OR if we do not expect the argument to have a value (`has_value` is false).  We only need to remove the element at position idx

Thus a1 becomes:
```
idx = 1
a1 = ['--foo', '--speed=88', '--location=home', '--enable-flux']
del a1[idx]
# idx is now  ['--foo', '--speed=88', '--location=home', '--enable-flux']
```

Now consider a2:
a2 = ['--foo', '--speed', '88', '--location=home', '--enable-flux']
In this case we actually want to remove `--speed` and `88` because `88` is the argument to speed but found in a seperate list value.

So therefore:
```
idx = 2
a2 = ['--foo', '--speed', '88', '--location=home', '--enable-flux']
del a2[idx]
del a2[idx]
# a2 is now ['--foo', '--speed', '--enable-flux']
```

I will try to add some comments on this to make it clearer.

Reply:
It would probably be a lot clearer if I change it to use slices instead. So I'll look into that too.

Reply:
I think I see what you mean. I was only thinking in terms of args using an "=" or that didn't have values at all. I was thinking how you would differentiate between something like `--boolean-flag-with-no-value --other-flag` and `--flag-with-value <value>`, but I guess the fact that you're replacing a specific flag that MUST have an `=` if it has a value (at least the replacement version of it, which is coming from us so can be controlled) then you know beforehand whether the flag will have a value and the issue goes away.